src/cargo/core/compiler/job_queue.rs

```diff
@@ -887,19 +886,14 @@ impl<'cfg> DrainState<'cfg> {
     fn handle_error(
         &self,
         shell: &mut Shell,
-        err_state: &mut Option<anyhow::Error>,
+        err_state: &mut ErrorsDuringDrain,
         new_err: anyhow::Error,
     ) {
-        if err_state.is_some() {
-            // Already encountered one error.
-            log::warn!("{:?}", new_err);
-        } else if !self.active.is_empty() {
-            crate::display_error(&new_err, shell);
+        crate::display_error(&new_err, shell);
```

Comment:
I don't fully understand why, but this breaks [`build::close_output` in the test suite](https://github.com/rust-lang/cargo/blob/86376c8dd41378144a7393fbc744102a36cdc008/tests/testsuite/build.rs#L5564-L5686).

The failure looks like:

```console
---- build::close_output stdout ----
thread 'build::close_output' panicked at 'called `Result::unwrap()` on an `Err` value: Expected lines did not match (ignoring order):
0   100002     Compiling foo v0.1.0 (/home/runner/work/cargo/cargo/target/tmp/cit/t231/foo)
1   0     hello stderr!
2   0     error: Broken pipe (os error 32)
3   99999  warning: build failed, waiting for other jobs to finish...
4   0     error: Broken pipe (os error 32)
    5    +error: Broken pipe (os error 32)
    6    +error: Broken pipe (os error 32)
    7    +error: Broken pipe (os error 32)
    8    +error: Broken pipe (os error 32)
    9    +error: Broken pipe (os error 32)
    10   +error: Broken pipe (os error 32)
    11   +error: Broken pipe (os error 32)
    12   +error: Broken pipe (os error 32)
    13   +error: Broken pipe (os error 32)
    14   +error: Broken pipe (os error 32)
    15   +error: Broken pipe (os error 32)
    16   +error: Broken pipe (os error 32)
    17   +error: Broken pipe (os error 32)
    18   +error: Broken pipe (os error 32)
    19   +error: Broken pipe (os error 32)
    20   +error: Broken pipe (os error 32)
[...]
    99989 +error: Broken pipe (os error 32)
    99990 +error: Broken pipe (os error 32)
    99991 +error: Broken pipe (os error 32)
    99992 +error: Broken pipe (os error 32)
    99993 +error: Broken pipe (os error 32)
    99994 +error: Broken pipe (os error 32)
    99995 +error: Broken pipe (os error 32)
    99996 +error: Broken pipe (os error 32)
    99997 +error: Broken pipe (os error 32)
    99998 +error: Broken pipe (os error 32)
    99999 +error: Broken pipe (os error 32)
    100000 +error: Broken pipe (os error 32)
    100001 +error: Broken pipe (os error 32)
    100002 +error: Broken pipe (os error 32)

', tests/testsuite/build.rs:5680:6
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    build::close_output

test result: FAILED. 2298 passed; 1 failed; 6 ignored; 0 measured; 0 filtered out; finished in 506.90s
```

The test involves a proc macro where we close its stdout and then execute `println!` 100,000 times from the macro. I believe println is supposed to panic on io error so I am confused how the macro keeps going and prints 99,999 more times after the first println has failed.

The test passes after the following diff:

```diff
-         crate::display_error(&new_err, shell);
+         if err_state.count > 1 {
+             log::warn!("{:?}", new_err);
+         } else {
+             crate::display_error(&new_err, shell);
+         }
```

Reply:
I believe the reason for this is that the main loop here gets one-line-at-a-time from the proc-macro, which is 100k prints. Each print fails because of a broken pipe, so Cargo tries to handle the error 100k times.

Short of specifically handling broken pipe errors here I'm not entirely sure what to do. We don't want to drop errors on the floor, especially with your future PR of continuing on error, so they need to be handled somehow I think?

Reply:
I'm a little concerned specifically about this change (and I realize it is the crux of this PR). It was pretty intentional to not show any errors after the first due to scenarios like this, where Cargo could just spew a bunch of error messages, but only one is really relevant.

Perhaps it is possible to put errors into two categories, one which is "important" and should be shown no matter what, and the other where it should only be shown if no other error has been shown? I think the only error that fits the former category is [this one](https://github.com/rust-lang/cargo/blob/5f611afe28f73bf1b5fdc71a6df53c51f2643fc9/src/cargo/core/compiler/job_queue.rs#L685), which is the "could not compile foo…" error. The rest of the errors are all things like failed to write to stdout or other internal errors that aren't too critical to display if something else has gone wrong. Would it be possible to wrap that error up in such a way that it will only show duplicates for that?

That will still result in more errors being display (imagine there are 2 crates building in parallel that both fail), but I think that's probably fine (and maybe even preferred?).

There still would be some risk that some relevant error would get elided, but at the moment I don't see any that really matter (and wouldn't be any worse than it is today).

Reply:
I have update the PR to revert to `log::warn!` instead of `crate::display_error` for errors after the first, except for the `Message::Finish` error that you flagged as important.